mut/index/SnootyManifest.py

```diff
@@ -0,0 +1,215 @@
+            str_list = []
+            for f in first:
+                str_list.append(f.value)
+            return ' '.join(str_list)
+        # Cowardly give up and just don't provide a preview.
+        else:
+            return None
+
+        
+    def findMetadata(self):
+        logger("\tFinding metadata")
+        robots = True
+        keywords = None
+        description = None
+
+        jsonpath_expr = parse('$..children[?(@.name==\'meta\')]..options')
+        results = jsonpath_expr.find(self.tree)
+        if results:
+            results = results[0].value
+            if "robots" in results and results["robots"] == "None":
+                robots = False
+            if "keywords" in results:
+                keywords = results["keywords"]
+            if "description" in results:
+                description = results["description"]
+
+        return robots, keywords, description
+
+    def get_noindex(self) -> bool:
+        # TODO: determine what the index / noindex rules should be
+        # with Product (DOP-)
+        logger("\tDetermining indexability")
+        noindex = False
+        reasons: List[str] = []
+        # If :robots: None in metadata, do not index
+        if self.robots == False:
+            noindex = True
+            reasons.append("robots=None in meta directive")
+
+        # If page has no title, do not index.
+        if self.title == None:
+            noindex = True
+            reasons.append("This page has 0 headings, not even the H1")
+
+        logger("\tnoindex: {}".format(noindex))
+        return noindex, reasons
+
+    def export(self) -> Optional[Dict[str, Any]]:
+        '''Generate the manifest dictionary entry from the AST source.'''
+
+        if self.noindex:
+            logger("\tRefusing to index {} because: {}".format(self.slug, ' '.join(self.reasons)))
+            return
+
+        document = {
+            "slug": self.slug,
+            "title": self.title,
+            "headings": self.headings,
+            "paragraphs": self.paragraphs,
+            "code": self.code,
+            "preview": self.preview,
+            "tags": self.keywords,
+        }
+
+        return document
+
+class Manifest:
+    '''Manifest to provide to Atlas search.'''
+    def __init__(self, url: str, includeInGlobalSearch: bool) -> None:
+        self.url = url
+        self.globally = includeInGlobalSearch
+        self.documents: List[Dict[str, Any]] = [] 
+
+    def add_document(self, document: Dict[str, Any]) -> None:
+        '''Add a document to the manifest'''
+        if not document == None:
+            self.documents.append(document)
+    
+    def export(self) -> str:
+        '''Return the manifest as json.'''
+        manifest = {
+            "url": self.url,
+            "includeInGlobalSearch": self.globally,
+            "documents": self.documents
+        }
+        return dumps(manifest, indent=4)
+
+def process_snooty_manifest_bson(path: str) -> Any:
```

Comment:
I'm increasingly a fan of using pathlib.Path whenever possible.

Bonus: then you can just do `data = decode_all(path.read_bytes())`

Reply:
I think I used that in the tests, but somehow didn't port it over here.